### DISCUSSION TITLE
Question: Cancellation

### DISCUSSION THREAD

**Discussion by ModernRonin** [TheCodeTraveler/AsyncAwaitBestPractices#59] [Q&A] [answered]
Hi,

first of all thanks for a very useful library :-)

My question is: how would I make an AsyncCommand cancellable? Scenario: cmd is bound to a long running operation, say a big download, that the user should be able to cancel via another command.

I can't seem to find any support for cancellation, but maybe I'm overlooking something?

Cheers & thanks,
MR

**Comment by johnthiriet** [comment, marked as answer]
Hello,

This might not be the best place to ask a question as it is not an issue per say but here is what you can do in pseudo code.

```
class ViewModel 
{
  CancellationTokenSource _cts;
  AsyncCommand StartCommand(async () => {
   try  {
     _cts?.Cancel();
    _cts?.Dispose();
    _cts = new CancellationTokenSource();
    StartYourProcessThatHasCancellationAsync(_cts.Token);
   }
   finally {
     _cts?.Dispose();
    _cts = null;
   }
  }
  
  AsyncCommand StopCommand(() => {
     _cts?.Cancel();
  });
}
```
